### ISSUE TITLE
Nunit3 runner can't run combination of Nunit2 and 3 tests?

### ISSUE THREAD

**Issue by netmailgopi** [nunit/nunit#2599]
So I have a mammoth solution full of about 2.5k tests all targetting Nunit 2.6.4. I'm in the process of migrating them to target Nunit3 (3.9). I migrated a bunch of projects, made some changes to overcome the breaking changes. When I run the build and test in TeamCity with Nunit console runner 3.4.1 I don't see any tests getting picked or run. The NUnit step is running for over 30m and no tests are getting identified. The Nunit2.6.4 runner runs all tests in about 8m. 

I also tried adding NUnit.Extension.NUnitV2Driver nuget package but it didn't change anything.

I see this same block getting repeated for each of the projects.

[11:52:17]Starting: C:\NUnit.ConsoleRunner.3.4.1\tools\nunit3-console.exe D:\BuildAgent\temp\buildTmp\pXqRm6ESTWkQ1qWrFNGRrkL3w5uV0xrV.nunit --result=D:\BuildAgent\temp\buildTmp\pXqRm6ESTWkQ1qWrFNGRrkL3w5uV0xrV.nunit.xml --noheader --where cat!=Integration&&cat!=Performance&&cat!=Staging --framework=net-4.0 --explore=D:\BuildAgent\temp\buildTmp\Ejgzmelp2EGVC9x2cNr3AMOaOg6GupWp_all.tests
[11:52:17]in directory: D:\BuildAgent\work\4d26b68a6ce28083\Projects\Services.Account.Web.Tests\bin\Staging
[11:52:17]Runtime Environment
[11:52:17]   OS Version: Microsoft Windows NT 10.0.14393.0
[11:52:17]  CLR Version: 4.0.30319.42000
[11:52:17]
[11:52:17]Test Files
[11:52:17]    D:\BuildAgent\temp\buildTmp\pXqRm6ESTWkQ1qWrFNGRrkL3w5uV0xrV.nunit
[11:52:17]
[11:52:17]Test Filters
[11:52:17]    Where: cat!=Integration&&cat!=Performance&&cat!=Staging
[11:52:17]
[11:52:47]Results (nunit3) saved as D:\BuildAgent\temp\buildTmp\Ejgzmelp2EGVC9x2cNr3AMOaOg6GupWp_all.tests
[11:52:47]Process exited with code 0

[12:00:24]Test Run Summary
[12:00:24]  Overall result: Failed
[12:00:24]  Test Count: 0, Passed: 0, Failed: 0, Inconclusive: 0, Skipped: 0
[12:00:24]  Start time: 2017-12-06 17:59:56Z
[12:00:24]    End time: 2017-12-06 18:00:26Z
[12:00:24]    Duration: 30.108 seconds

**Comment by CharliePoole** [comment]
Can you post one of the result files?

**Comment by netmailgopi** [comment]
Thanks for the quick response. I looked at one of the files and here's the content:

<?xml version="1.0" encoding="utf-8" standalone="no"?>
<test-run id="2" testcasecount="1" result="Failed" label="Error" total="1" passed="0" failed="1" inconclusive="0" skipped="0" asserts="0">
  <test-suite type="Assembly" id="1" name="cSlH58ZIFslfwUiaAdOjdH7ZhD11N9I7.nunit" fullname="D:\BuildAgent\temp\buildTmp\cSlH58ZIFslfwUiaAdOjdH7ZhD11N9I7.nunit" runstate="NotRunnable" testcasecount="1" result="Failed" label="Error" start-time="2017-12-06 19:51:44Z" end-time="2017-12-06 19:51:44Z" duration="0.001" total="1" passed="0" failed="1" inconclusive="0" skipped="0" asserts="0">
    <failure>
      <message><![CDATA[Unable to acquire remote process agent]]></message>
      <stack-trace><![CDATA[   at NUnit.Engine.Runners.ProcessRunner.CreateAgentAndRunner()
   at NUnit.Engine.Runners.ProcessRunner.Explore(TestFilter filter)]]></stack-trace>
    </failure>
  </test-suite>
</test-run>

I did have --inprocess arguement previously to get over this error but I had to remove because it gave me this error. 

[11:17:26] :	 [Step 3/4] NUnit.Engine.NUnitEngineException: Cannot run tests in process - a 32 bit process is required.
[11:17:26] :	 [Step 3/4]    at NUnit.Engine.Runners.MasterTestRunner.InitializePackage()
[11:17:26] :	 [Step 3/4]    at NUnit.Engine.TestEngine.GetRunner(TestPackage package)
[11:17:26] :	 [Step 3/4]    at NUnit.ConsoleRunner.ConsoleRunner.ExploreTests(TestPackage package, TestFilter filter)
[11:17:26] :	 [Step 3/4]    at NUnit.ConsoleRunner.Program.Main(String[] args)
[11:17:26]W:	 [Step 3/4] Process exited with code -100

I also don't want to use --inprocess as I want to leverage parallel tests with NUnit3 runner once the migration is complete.

**Comment by CharliePoole** [comment]
I think you may be running into a bug that we fixed. Can you switch to the latest version of the console runner?

**Comment by netmailgopi** [comment]
Sure, trying with 3.7

**Comment by netmailgopi** [comment]
I tried running with 3.7 console runner and I get "File Type not supported error"

`<?xml version="1.0" encoding="utf-8" standalone="no"?>
<test-run id="2" testcasecount="0">
  <test-suite type="Unknown" id="0-1" name="VGNlK2g61kIqYZskAvU2Q451C5XVepgu.nunit" fullname="D:\BuildAgent\temp\buildTmp\VGNlK2g61kIqYZskAvU2Q451C5XVepgu.nunit" testcasecount="0" runstate="NotRunnable">
    <properties>
      <property name="_SKIPREASON" value="File type is not supported" />
    </properties>
  </test-suite>
</test-run>`

**Comment by CharliePoole** [comment]
Apparently you haven't installed the nunit project loader extension.

**Comment by netmailgopi** [comment]
I have that nuget package installed for a project. Do you mean installing on the machine/Teamcity?

`PM> Install-Package NUnit.Extension.NUnitProjectLoader -Project Services.Account.Libraries.Tests
'NUnit.Extension.NUnitProjectLoader 3.6.0' already installed.`

**Comment by CharliePoole** [comment]
Installing the nuget extension for a project makes it available to the console runner installed in the same way. But if the machine running teamcity is not using a nuget package to access the console, then it won't have that extension. Extensions are installed against a specific runner/engine installation.

Solution is to use the same installation approach for the console/engine and any extensions.

**Comment by netmailgopi** [comment]
I rechecked everything again. We don't have NUnit 3 installed on our buildagents. TeamCity comes with Nunit2 packaged. We use the relative path .\Packages\..NUnit3consolerunner to pick and run the tests. I looked up and also added NUnit.Console package that bundles everything in it (https://www.nuget.org/packages/NUnit.Console) 

During the build I see the packages getting restored but the tests don't seem to get picked. I looked at the log files, File Type not supported again. It's possible that the V2 tests are not picked for some reason, wondering why V3 tests are not getting captured.

**Comment by CharliePoole** [comment]
TeamCity does an Explore of the tests using the `.nunit` file it creates. Can you ascertain which copy of the console runner is being used for that explore operation? Note that extensions only operate with the copy of the engine against which they are installed.

**Comment by netmailgopi** [comment]
It does use the runner 3.7 that I've specified. About the copy of engine they're installed against, how do map the extensions to the runner? I don't remember doing anything explicitly. 

`[10:52:26]Step 4/5: NUnit Tests (NUnit) (1m:21s)
[10:52:26]The TeamCity NUnit runner requires the following NUnit extensions to be installed: NUnit.Engine.Listeners.TeamCityEventListener, NUnit.Engine.Services.ProjectLoaders.NUnitProjectLoader. Please follow our instructions at: https://confluence.jetbrains.com/display/TCDL/NUnit#NUnit-NUnit3Extensions
[10:52:27]Starting: D:\BuildAgent\work\4d26b68a6ce28083\Packages\NUnit.ConsoleRunner.3.7.0\tools\nunit3-console.exe D:\BuildAgent\temp\buildTmp\6GhXbJ5CZ1Ez9Z9GOimEUAK2KLgsZY84.nunit --result=D:\BuildAgent\temp\buildTmp\6GhXbJ5CZ1Ez9Z9GOimEUAK2KLgsZY84.nunit.xml --noheader --where cat!=Integration&&cat!=Performance&&cat!=Staging --framework=net-4.0 --explore=D:\BuildAgent\temp\buildTmp\9fJ8e6vnlt49oiEsTX1XGigYUEOCCDw1_all.tests`

**Comment by CharliePoole** [comment]
Speaking from the user point of view:
 * If you installed the engine/runner using the msi, it uses the extensions that are packaged in the same msi.
 * If you installed using NuGet, it uses any extensions installed in the same project via nuget. 
 * If you installed with Chocolatey, it uses extensions that are installed with chocolatey.

From the internal perspective...
 * Each copy of the engine looks in it's own directory for one or more files of type `.addins` and follows links contained in the file to find extensions. The `.addins` file may reference dlls, directories or other `.addins` files recursively.

**Comment by netmailgopi** [comment]
Thanks for clarifying the addins part. I was able to deep dive and get some observations. I think the problem lies with the addins. 

So my NUnit.ConsoleRunner.3.7.0/tools has just one nunit.nuget.addins and nothing else. However, I have a  NUnit.Console.3.7.0 directory next to the console runner that has the only console.nupkg 

My packages.config files for the test projects look like,
`<package id="NUnit" version="2.6.4" targetFramework="net45" /> (Either 2.6.4 or 3.9.0) 
 <package id="NUnit.Console" version="3.7.0" targetFramework="net45" />
 <package id="NUnit.ConsoleRunner" version="3.7.0" targetFramework="net45" />`

How do I restore the addins into the runner directory? 

But if I manually run the test dll file with the console runner I'm able to discover and run the tests. 
`PS> nunit3-console.exe  /framework:net-4.0 ..\Services.Account.Web.Tests\bin\Debug\Services.Account.Web.Tests.dll` 

When TeamCity tries the explore I think it is not able to discover the tests, as the addins are not present? I see this line in the TeamCity log if that matters,
`[18:30:06]The TeamCity NUnit runner requires the following NUnit extensions to be installed: NUnit.Engine.Listeners.TeamCityEventListener, NUnit.Engine.Services.ProjectLoaders.NUnitProjectLoader. Please follow our instructions at: https://confluence.jetbrains.com/display/TCDL/NUnit#NUnit-NUnit3Extensions`

**Comment by netmailgopi** [comment]
I got it. The .addin file looks for addin packages and my Packages directory didn't have the NUnit,Extensions packages pulled. I copied them from some other project and pasted on the teamcity work directory. After this, Teamcity was able to discover and run tests. Having NUnit.Console alone in the packages.config did not pull all the packages. I had to explicitly list all of them to get it. Thanks for helping @CharliePoole Appreciate it.
